README.md
[![Build Status](https://github.com/marshallmcmullen/ebash/workflows/Pipeline/badge.svg?branch=master)](https://github.com/marshallmcmullen/ebash/actions?query=workflow%3APipeline+branch%3Amaster)

<p align="center">
    <img alt="Bash" src="https://raw.githubusercontent.com/odb/official-bash-logo/master/assets/Logos/Identity/PNG/BASH_logo-transparent-bg-color.png">
</p>

# ebash Guide

## Overview

Bash is the ideal language of choice for writing low-level shell scripts and tools requiring direct shell access
invoking simple shell commands on a system for a number of reasons:

1. Prolific and already installed on nearly every *NIX machine out of the box which minimizes dependencies and
   installation complexity and size.
2. Extremely lightweight with very low memory and CPU requirements suitable for appliances and embedded systems.
3. Ideal for tasks involving running lots of shell commands as it is simpler than in higher level languages. This is,
   after all, bash's primary native role. As such, it does not require complex objects or frameworks for executing
   or interacting with shell commands.

That said, because bash is a lower level language, it lacks some of the features and more advanced data structures
typically found in higher level languages. As such, the goal of ebash is to provide a layer of robustness and more
advanced features on top of vanilla bash.

## Compatibility

Ebash aims for extreme compatibility with the intention that it "just works" on just about any *NIX OS. While it is not
feasible to test ebash on _every_ OS/Distro, we nonetheless test it many popular combinations on every single build. You
can view the build status any time using the build batch at the top of this README or by going here:
[Builds](https://github.com/marshallmcmullen/ebash/actions?query=workflow%3APipeline+branch%3Amaster).

Here is a table showing all the currently test and validated OS/Distro
combinations:

| OS    | Disto  | Version |
| ----- | -------| ------- |
| Linux | Alpine | 3.11    |
| Linux | Alpine | 3.12    |
| Linux | Arch   | rolling |
| Linux | CentOS | 7       |
| Linux | CentOS | 8       |
| Linux | Debian | 10      |
| Linux | Debian | 9       |
| Linux | Fedora | 32      |
| Linux | Fedora | 33      |
| Linux | Gentoo | rolling |
| Linux | Ubuntu | 16.04   |
| Linux | Ubuntu | 18.04   |
| MacOS | N/A    |  11.0   |
| MacOS | N/A    | 10.15   |

## Why ebash?

### Implicit Error Detection

Bash is a strange beast.  It can be an interactive shell, in which case you’d like to be notified of failures but not
for them to be catastrophic.  And when you write scripts, it’s easy to be torn about what you want.  Sometimes you want
failures to be ignored, and other times you want to be very sure that a command doesn’t fail.

The standard bash answer to this question is that you should always check for errors if you care about them.  But
sometimes commands fail that you don’t expect to fail.  What happens when you’ve left out a test in that case because
that could never happen, right?  Or perhaps you have taken extreme care to check for errors in all possible cases.  Now
your code is wordy and more difficult to follow.

With ebash, we have decided to make scripts detect errors implicitly.  If an error occurs, you’ll detect it and your
script will blow up.  You can specifically ignore such failures when you know it’s safe, but the default is to make the
failure as obvious as possible.

This is in opposition to all bash defaults.  And, frankly, it’s a bit difficult to force bash into really blowing up on
all failures.  But we do our best with a multi-layered approach.  As long as you use the typical ebash idioms and follow
a small set of styles we prescribe, we believe that any time an error occurs in your ebash code, you’ll hear about it.

First, we use bash’s error trap functionality.  This behaves just like `set -e`, except that we get to choose what runs
when an error occurs.  Basically, we call the die function on any error.  This causes a stack trace to be generated
which will have `[UnhandledError]` in its message.  Second, we ensure that the error trap is inherited by all subshells,
because sometimes you need them, and often it’s difficult to know where there will actually be a subshell.

But sometimes bash really wants to ignore a return code.  For instance, if command substitution isn’t part of the first
command on a line, as in both of the following lines of code, bash will ignore any return code of `cmd`.

    echo "$(cmd)"
    local var=$(cmd)

But ebash enhances bash’s error detection to catch this case.  The error trap is inherited by the shell inside the
command substitution, and when it calls die, it will actually send a signal to its parent shell.  When the parent shell
receives that signal, it generates a stack trace and blows up, which is exactly what we were looking for.

However, despite the best efforts of ebash, bash allows you to shoot yourself in the foot in subtle ways that mask error
handling.  The final component of our efforts to make errors implicitly detected are some styles to avoid.

### Using || or && masks errors

Typically, you can count on bash’s built in error detection (i.e. `set -e`, which we enable by default) to blow up if a
command you run exits with a non-zero exit code.  Sometimes that’s not what you want.  For one example, see `grep`
returns a bad exit code if it has no matches.

But using `||` has a more insidious effect when used around a bash function call, and for that reason we avoid it
whenever possible on bash functions.  Consider:

    func()
    {
        command1
        command2
    }
    func


When this code is executed, if `command1` fails, then the script will blow up immediately and will not execute
`command2`.  In ebash, this is what we want and expect.  But, given the same `func`, the following command works
differently:

    func && echo "success" || echo "failure"

Because `func` is part of a command list containing either `&&` or `||`, bash turns off its error detection behavior for
all code that is called, not just the code written on this line.  So what happens in this case is that `func`runs.  It
calls `command1`, which fails.  But since bash’s error detection behavior is off, it goes on to call `command2` which
passes and thus the overall result of `func` is success.

As you can see, this can dramatically affect the way code runs, and we’ve come across large issues that were masked by
this behavior.  So our standard on these is to only use `||` or `&&` on simple statements including only bash builtins
or executables.

If you need to ignore errors in bash functions that you call, instead use `tryrc` or `try` / `catch`. They make sure
that error detection is enabled for all called code.

### Calling bash functions in the if or loop conditional masks errors

This issue is similar to Using `||` or `&&` masks errors.  If you assume the same `func` as in that item, issues called
in `func` will also be masked by either of the following types of code:

    if func; then ; something ; fi

    while func; do ; something ; done

That is to say, any errors in anything `func` calls will be suppressed, so it is likely that the behavior and return
value of `func` isn’t want you want expect it to be.  One way around this is to use `tryrc`.  The following is
equivalent to the above if code, except that it causes `func` to detect errors properly.

    $(tryrc func)
    if [[ ${rc} -eq 0 ]] ; then
        something
    fi

### Use try / catch to handle errors

Since ebash forces you to be explicit about when you’d like to ignore errors in your bash script, it also gives you ways
to do that.  First off is the `try` / `catch` construct.  This looks frighteningly like the C++ construct.

    try
    {
        cmd1
        cmd2
    }
    catch
    {
       echo "Caught return code $?"
    }

It also behaves a lot like you’d expect from familiarity with other languages.  Code in the try block is executed until
the end of the block or until an error is encountered.  At the point an error is detected, no further code in the block
will execute.  So if `cmd1` fails in the above block, `cmd2` will never be executed.

The catch block is only executed if a failure is detected in the try block.  So if either `cmd1` or `cmd2` fail, the
catch block will be executed.  At the beginning of the catch block, the value of `$?` will be the return code of the
failing command.

The most difficult caveat of using `try` and `catch` in ebash is that the code in the `try` block is its own subshell.
This means that variable assignments that you make inside this block will not be seen outside.  It’s impossible to set a
variable in a parent shell from its subshell.  Note that this is only true of `try`.  `Catch` executes in the same shell
as surrounding code.

One workaround is to write data to a file inside the block and read it outside.  Another alternative is `tryrc`.

### Use tryrc to mask errors in simple commands

Sometimes you just need to run a single command and then perform different behavior depending on its return code.  For
external commands, the bash if statement or shortcutting logical operators (`||` and `&&`) handle this nicely, but if
you’re calling functions it is a bad idea to use them in this way.

`Tryrc` is the ebash solution to this issue.  It runs a function (or external command, basically any single statement
bash can execute) and captures its return code for later examination.  Its simplest usage looks like this:

    $(tryrc foo)
    if [[ ${rc} -eq 0 ]] ; then
        # Normal case
    else
        # Error case
    fi

The `tryrc` call will execute the foo function or command, and save its return code in a new local variable (created by
`tryrc`) named `rc`.  Error detection in code called by `foo` is still enabled, but regardless of the return code, the
`tryrc` statement will not fail.  Then you’re able to handle the return code as you like.

`tryrc` is also capable of capturing the stdout and/or stderr of the executed command into variables that you specify
during at invocation time.  Here’s a more complicated usage:

    $(tryrc -r=cmd_rc -o=cmd_out -e=cmd_err cmd)

This invocation will execute cmd and create three local variables.  `Cmd_rc` will contain the return code of `cmd`.
`Cmd_out` and `cmd_err` will contain stdout and stderr, respectively.  This gives you the functionality of command
substitution along with the ability to choose what happens with errors.

This functionality is important, because it’s difficult to use redirection with commands that are executed in `tryrc`.
If you need use redirection in way that can’t be solved by something like the following, it’s best to use `try` /
`catch` instead of `tryrc`.

    $(tryrc -o=out -e=err cmd)
    echo "${out}"
    echo "${err}" >&2

Note that if `cmd` produces interleaved output on stdout and stderr, this will change the ordering of the output.

### Handling commands that fail or hang

The ebash `eretry` function can be used to run other commands that you expect to sometimes fail, but from which you
eventually expect a positive response.  Suppose I have a tool named `foo` that frequently fails in flaky ways.  When I
call it with no options, `eretry` will run `foo` until it passes, with up to 5 attempts.

    eretry foo args

`Eretry` can also simply keep retrying as many times as it takes to reach a certain time period that you specify.  If
I’m pretty sure `foo` will pass within a total of 5 seconds, I can call it like this:

    eretry -T=5s foo args

`Eretry` will keep running `foo` until 5 seconds have elapsed.  If it hasn’t passed yet at that point, `foo` will be
killed and `eretry` will return a failing exit code.  Further, you can restrict how long each attempt at running `foo`
will take.  So the following command will run for up to a total (-T) of 5 seconds, allowing individual attempts to take
as much as 1 second (-t)

    eretry -T=5s -t=1s foo args

Or, if you simply need to run a command once, but make sure it doesn’t hang for too long, you can use `etimeout`.  In
the following call, `etimeout` allow `foo` to run for one minute, but if it hasn’t completed at that point, `foo` will
be killed and an error code will be returned.

    etimeout -t=60s foo

### `grep` returns a bad exit code if it has no matches

When you use `grep`, be aware that it will return a bad exit code if it finds no matches.  And when it does, the ebash
error detection code will blow your script up to let you know that there was an error.  But sometimes that isn’t what
you want.

    var=$(some_command | grep pattern)

The above will work in your testing, but suppose there isn’t a match.  In that case, your script will blow up with a
stack trace.  If it’s okay for there to be an error, just use `|| true` to mask it.

    var=$(some_command | grep pattern || true)

### Unset variables don’t invoke the error trap

They do, however, exit immediately.  This is unfortunately how bash’s `nounset` option works.  I can’t imagine a very
good explanation as to why bash wouldn’t invoke its error trap if there is one, but that’s how life is.

As long as you handle your cleanup in an `EXIT` trap, which is what `trap_add` does by default, it’ll be fine.

## Common Logging and Output Styles ### A standard, styleable output format

ebash provides a few standard functions intended for logging output to the screen.  `Einfo` is intended for standard
flow information in the typical case.  `Ewarn` indicate non-fatal things that might warrant user attention.  `Eerror`
are typically used for failure cases.  All of these produce their output on stderr.

These commands are invoked in the same way as you might call `echo`.  Here’s a script that uses them.

    #!/usr/bin/env bash

    : ${EBASH:=/home/modell/sf/ebash}
    source ${EBASH}/ebash.sh || exit 1

    einfo "einfo"
    ewarn "ewarn"
    eerror "eerror"

I’m not going to try to reproduce color in this guide, but by default the
output looks like this, with green color for info-level messages, yellow for
warning, and red for errors.

    >> einfo
    >> ewarn
    >> eerror

But the output can be modified in a few ways.  One way to customize it is to set `EFUNCS_COLOR=0`, which turns of color
for this and all ebash tools.  Another is that you can request timestamps.

    > EMSG_PREFIX=time ~/ebash_guide
    [Nov 12 13:31:16] einfo
    [Nov 12 13:31:16] ewarn
    [Nov 12 13:31:16] eerror

Or, to request all information that ebash will produce, you can set `EMSG_PREFIX` to all.

    > EMSG_PREFIX=all ./ebash_guide
    [Nov 12 13:24:19|INFO|ebash_guide:6:main] einfo
    [Nov 12 13:24:19|WARN|ebash_guide:7:main] ewarn
    [Nov 12 13:24:19|ERROR|ebash_guide:8:main] eerror

Here you can see the timestamp, log level, function name, line number, and filename of the code that generated the
message.  There’s more information on emsg customization in this other guide.

Another built-in output tool is `ebanner`.  It helps you separate high-level sections of the output of a process.  The
bars above and below your text will span the entire width of the terminal.

    > ebanner "Here’s a banner"
    +--------------------------------------+
    |
    | Here's a banner
    |
    +--------------------------------------+

### Use `edebug` and `lval` for more verbose logging

Despite the best of intentions, everyone writing bash code eventually needs to starting printing things out to figure
out what on earth is happening.  For user-level logging, ebash includes `einfo`, `ewarn`, and `eerror` commands that
present messages of varying importance in a consistent manner.

But back to debugging, we also have an `edebug` (which we pronounce "ee-debug") command that is a little different.  By
default, the `edebug` logging "level" is hidden.  So by default, this statement won’t produce any output:

    edebug "foo just borked rc=${rc}"

But you can activate the output from these `edebug` statements either wholesale or selectively by setting an environment
variable.  Setting `EDEBUG=1` will turn on all `edebug` output.  We use it pretty pervasively, so that might be a lot.
Instead of turning everything on, you can turn on `edebug` just for code in certain files or functions.  For example,
using `EDEBUG="dtest dmake"` will turn on debugging for any `edebug` statements in the `dtest` or `dmake` scripts.

What if you want to send the entire output of another tool to `edebug`?  You can do that, too.

    cmd |& edebug

The value of `EDEBUG` is actually a space-separated list of terms.  If any of those terms match the filename (just
basename) or the name of the function that contains an `edebug` statement, it will generate output.

Because we frequently found ourselves typing things like `rc=${rc}` in our `edebug` and other logging statements, we
also created a function called `lval` to help with this.  If you instead use `$(lval rc)`, it will produce the same
output.

It also prints the output in a format that makes it clear what is in your variable, even when it contains whitespace.
And it knows about bash types like arrays and associative arrays and handles them appropriately.  You might use it like
this.  You can, of course, use it with any tool that outputs an arbitrary string.

    declare -A aa
    aa[foo]=1
    aa[bar]=2
    local var="hello world"
    array[0]=alpha
    array[1]=beta
    echo "$(lval aa var array)"
    # Produces:
    # aa=([bar]="2" [foo]="1" ) var="hello world" array=("alpha" "beta")

### Debugging beyond logging: `ETRACE`

But maybe you’ve looked at all the debugging output you can find and you still need more information about what is going

changelog.md
# ebash 1.4

    - Moved all the eprompt code out of emsg.sh into its own eprompt.sh
      module. There were no unit tests for this code, so a new eprompt.etest
      was created.

    - Modified chroot unittests to only create a chroot once. The tests
      then all re-use that single chroot instead of each test creating
      its own copy. This shaves 20 minutes off of our test time.

    - New dialog.sh module provides a simple interface for using dialog
      ncurses interface (http://invisible-island.net/dialog/) within bash.
      This module provides some convenience wrappers to make it easier to
      obtain the output from dialog and its return code without without
      causing the ERR trap to be invoked via the "eval command invocation
      string" idiom. Also provides convenience methods dialog_read,
      dialog_prompt, dialog_prgbox, dialog_prompt_username_password and a
      very large set of unit tests around this new module. At this time
      there are some bugs with dialog on older Ubuntu 12.04 and also on
      OSX so the module is excluded from those OSes.

    - Added optional --uppercase flag to ebanner to emit the keys in all
      uppercase for greater uniformity if the caller desires it.

    - Add new --strip-components/--strip flag to archive_extract to make
      it easier to use in place of tar when you need to strip off leading
      path components.

    - Add --align option to eprogress with valid values "left" and "right".
      The new default is "right" to keep the progress tickers aligned with
      where the "[ ok ]" shows up aftewards.

# ebash 1.3

    - New configuration file reading and writing tools for INI-style
      configuration files.  (see share/conf.sh for more info)

    - Added pkg subsystem with tools for installing and uninstalling
      packages as well as determining whether they're installed in a safe
      way across Linux distros.

    - Added --file option to eprogress

    - Added --hexdump option to assert_eq

    - Added -n option to einfo, ewarn, edebug that prevents them from
      generating a newline at the end of the message.  Note that this must
      be held in the first argument to the einfo/ewarn/edebug function.

    - Corrected substantial defects in the existing assert functions and
      added unit tests for them.

    - assert_empty and assert_not_empty now accept strings and blows up if
      any of the strings do not meet their criteria.  Previously they
      expected the name of a variable.  Assert_var_empty and
      assert_var_not_empty have been added to handle that use case.

    - Created a new opt_parse function that contains all of the argument
      handling functionality that `declare_args` used to while also being
      explicit instead of implicit about option handling.  Supports both short
      and long options.  Supports "golfing" of short options (i.e. -e -x and
      -ex are equivalent).  Detects errors on the command line of calling code
      that would've previously been ignored such as passing an unsupported
      option to a command.

    - New opt_forward function that helps forward options passed into one other
      functions that accept the same options. Useful when you have an
      "internal" function that handles most of the work that is done by a slim
      calling function.

    - Removed declare_args in favor of opt_parse.  It is syntax-compatible as
      long as you didn't use options.  When using options, they must now be
      specified but we prefer this as it is more explicit.

    - Fixed a bug that caused die not to signal its parent when aborting.  It
      is through this mechanism that we intend to catch errors that happen in
      command substitution shells, so this may unearth some errors that were
      previously undetected.  If you want to disable this for a particular
      shell, you can call disable_die_parent in it.

    - New functions to assist in ascertaining your position in the process
      hierarchy.  Process_tree already existed, but process_children,
      process_parent, and process_ancestors are new.

    - ekill now supports an option (--kill-after) to send the initial signal,
      plus send sigkill later if the process doesn't die, and similar
      functionality has been added to other functions that kill such as
      ekilltree.

    - Rather than depth-first iteration, ekilltree now kills the children of a
      process all at the same time.

    - Die now only kills itself via the terminating signal it caught when that
      signal is one of the TTY signals.  Doing that on other signals such as
      sigterm can cause bash not to execute exit traps.

    - The daemon utilities no longer support a post_start hook but do support a
      post_mount hook.  That ability exists mostly to aid in testing the daemon
      utilities, but perhaps it is also useful elsewhere.  Better logging
      should be produced by the daemon ebash functions into the daemon's
      log file if you request one.

    - Chroot_kill now defaults to sigterm rather than sigkill.

    - Exceptions, unhandled errors, and caught signals now report the pid of
      the running process and the command running at the time of occurrence.

    - Renamed all internal "global" variables to begin with "\_\_BU\_".  It's
      often shorter than what we were previously using which is nice, but the
      big thing here is that we can tell people to not use those.

    - Etest now runs under Mac OS, given appropriate GNU toolchains installed
      with names prefixed with "g" (which seems to be the default for both
      homebrew and macports).

    - Ecolor now allows for 0 or more arguments which can be modifiers or
      foreground colors. No longer support "dimCOLOR".  Instead use "dim
      color". Ecolor now also supports setting background color by prefixing
      the color name with "b:"

    - Ecolor caches its output so that it doesn't have to spawn tput processes

network.sh
#!/usr/bin/env bash
#
# Copyright 2011-2016, SolidFire, Inc. All rights reserved.
#

#-----------------------------------------------------------------------------
# NETWORKING FUNCTIONS
#-----------------------------------------------------------------------------
valid_ip()
{
    $(opt_parse ip)
    local stat=1

    if [[ $ip =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$ ]]; then
        array_init ip "${ip}" "."
        [[ ${ip[0]} -le 255 && ${ip[1]} -le 255 && ${ip[2]} -le 255 && ${ip[3]} -le 255 ]]
        stat=$?
    fi
    return $stat
}

hostname_to_ip()
{
    $(opt_parse hostname)

    local output="$(host ${hostname} | grep ' has address ' || true)"

    [[ ${output} =~ " has address " ]] || { ewarn "Unable to resolve ${hostname}." ; return 1 ; }

    ip=$(echo ${output} | awk '{print $4}')

    valid_ip ${ip}
    echo ${ip}
    return 0
}

fully_qualify_hostname()
{
    local hostname=${1,,}
    argcheck hostname

    try
    {
        local output=$(host ${hostname})

        [[ ${output} =~ " has address " ]]
        local fqhostname=$(echo ${output} | awk '{print $1}')
        echo "${fqhostname,,}"
    }
    catch
    {
        ewarn "Unable to resolve ${hostname}"
        return 1
    }

    return 0
}

[[ ${__BU_OS} == Linux ]] || return 0

#-----------------------------------------------------------------------------
# Linux-specific networking functions
#-----------------------------------------------------------------------------

# Get the IPAddress currently bound to the requested interface (if any). It is
# not an error for an interface to be unbound so this function will not fail if
# no IPAddress is set on the interface. Instead it will simply return an empty
# string.
getipaddress()
{
    $(opt_parse iface)
    ip addr show "${iface}" | awk '/inet [0-9.\/]+ .*'${iface}'$/ { split($2, arr, "/"); print arr[1] }' || true
}

# Get the netmask (IPv4 dotted notation) currently set on the requested
# interface (if any). It is not an error for an interface to be unbound so this
# method will not fail if no Netmask has been set on an interface. Instead it
# will simply return an empty string.
getnetmask()
{
    $(opt_parse iface)
    local cidr=$(ip addr show "${iface}" | awk '/inet [0-9.\/]+ .*'${iface}'$/ { split($2, arr, "/"); print arr[2] }' || true)
    [[ -z "${cidr}" ]] && return 0

    cidr2netmask "${cidr}"
}

# Convert a netmask in IPv4 dotted notation into CIDR notation (e.g 255.255.255.0 => 24).
# Below is the official chart of all possible valid Netmasks in quad-dotted decimal notation
# with the associated CIDR value:
#
# { "255.255.255.255", 32 }, { "255.255.255.254", 31 }, { "255.255.255.252", 30 }, { "255.255.255.248", 29 },
# { "255.255.255.240", 28 }, { "255.255.255.224", 27 }, { "255.255.255.192", 26 }, { "255.255.255.128", 25 },
# { "255.255.255.0",   24 }, { "255.255.254.0",   23 }, { "255.255.252.0",   22 }, { "255.255.248.0",   21 },
# { "255.255.240.0",   20 }, { "255.255.224.0",   19 }, { "255.255.192.0",   18 }, { "255.255.128.0",   17 },
# { "255.255.0.0",     16 }, { "255.254.0.0",     15 }, { "255.252.0.0",     14 }, { "255.248.0.0",     13 },
# { "255.240.0.0",     12 }, { "255.224.0.0",     11 }, { "255.192.0.0",     10 }, { "255.128.0.0",      9 },
# { "255.0.0.0",        8 }, { "254.0.0.0",        7 }, { "252.0.0.0",        6 }, { "248.0.0.0",        5 },
# { "240.0.0.0",        4 }, { "224.0.0.0",        3 }, { "192.0.0.0",        2 }, { "128.0.0.0",        1 },
#
# From: https://forums.gentoo.org/viewtopic-t-888736-start-0.html
netmask2cidr ()
{
    # Assumes there's no "255." after a non-255 byte in the mask 
    set -- 0^^^128^192^224^240^248^252^254^ ${#1} ${1##*255.} 
    set -- $(( ($2 - ${#3})*2 )) ${1%%${3%%.*}*} 
    echo $(( $1 + (${#2}/4) ))
}

# Convert a netmask in CIDR notation to an IPv4 dotted notation (e.g. 24 => 255.255.255.0).
# This function takes input in the form of just a singular number (e.g. 24) and will echo to
# standard output the associated IPv4 dotted notation form of that netmask (e.g. 255.255.255.0).
#
# See comments in netmask2cidr for a table of all possible netmask/cidr mappings.
#
# From: https://forums.gentoo.org/viewtopic-t-888736-start-0.html
cidr2netmask ()
{
    # Number of args to shift, 255..255, first non-255 byte, zeroes
    set -- $(( 5 - ($1 / 8) )) 255 255 255 255 $(( (255 << (8 - ($1 % 8))) & 255 )) 0 0 0
    [ $1 -gt 1 ] && shift $1 || shift
    echo ${1-0}.${2-0}.${3-0}.${4-0}
}

# Get the broadcast address for the requested interface, if any. It is not an
# error for a network interface not to have a broadcast address associated with
# it (e.g. loopback interfaces). If no broadcast address is set this will just
# echo an empty string.
getbroadcast()
{
    $(opt_parse iface)
    ip addr show "${iface}" | awk '/inet [0-9.\/]+ brd .*'${iface}'$/ { print $4 }' || true
}

# Gets the default gateway that is currently in use, if any. It is not an
# error for there to be no gateway set. In that case this will simply echo an
# empty string.
getgateway()
{
    route -n | awk '/UG[ \t]/ { print $2 }' || true
}

# Compute the subnet given the current IPAddress (ip) and Netmask (nm). If either
# the provided IPAddress or Netmask is empty then we cannot compute the subnet.
# As it's not an error to have no IPAddress or Netmask assigned to an unbound
# interface, getsubnet will not fail in this case. The output will be an empty
# string and it will return 0.
getsubnet()
{
    $(opt_parse "?ip" "?nm")
    [[ -z "${ip}" || -z "${nm}" ]] && return 0

    IFS=. read -r i1 i2 i3 i4 <<< "${ip}"
    IFS=. read -r m1 m2 m3 m4 <<< "${nm}"

    printf "%d.%d.%d.%d" "$((i1 & m1))" "$(($i2 & m2))" "$((i3 & m3))" "$((i4 & m4))"
}

# Get the MTU that is currently set on a given interface.
getmtu()
{
    $(opt_parse iface)
    ip addr show "${iface}" | grep -Po 'mtu \K[\d.]+'
}

# Get list of network interfaces
get_network_interfaces()
{
    for iface in $(ls -1 /sys/class/net); do
        # Skip virtual devices, we only want physical
        [[ ! -e /sys/class/net/${iface}/device ]] && continue
        echo "${iface}"
    done | tr '\n' ' ' || true
}

# Get list network interfaces with specified "Supported Ports" query.
get_network_interfaces_with_port()
{
    local query="$1"
    local ifname port
    local results=()

    for ifname in $(get_network_interfaces); do
        port=$(ethtool ${ifname} | grep "Supported ports:" || true)
        [[ ${port} =~ "${query}" ]] && results+=( ${ifname} )
    done

    echo -n "${results[@]}"
}

# Get list of 1G network interfaces
get_network_interfaces_1g()
{
    get_network_interfaces_with_port "TP"
}

# Get list of 10G network interfaces
get_network_interfaces_10g()
{
    get_network_interfaces_with_port "FIBRE"
}

# Get the permanent MAC address for given ifname.
# NOTE: Do NOT use ethtool -P for this as that doesn't reliably
#       work on all cards since the firmware has to support it properly.
get_permanent_mac_address()
{
    $(opt_parse ifname)

    if [[ -e /sys/class/net/${ifname}/master ]]; then
        sed -n "/Slave Interface: ${ifname}/,/^$/p" /proc/net/bonding/$(basename $(readlink -f /sys/class/net/${ifname}/master)) \
            | grep "Permanent HW addr" \
            | sed -e "s/Permanent HW addr: //"
    else
        cat /sys/class/net/${ifname}/address
    fi
}

# Get the PCI device location for a given ifname
# NOTE: This is only useful for physical devices, such as eth0, eth1, etc.
get_network_pci_device()
{
    $(opt_parse ifname)

    (cd /sys/class/net/${ifname}/device; basename $(pwd -P))
}

# Export ethernet device names in the form ETH_1G_0=eth0, etc.
export_network_interface_names()
{
    local idx=0
    local ifname

    for ifname in $(get_network_interfaces_10g); do
        eval "ETH_10G_${idx}=${ifname}"
        (( idx+=1 ))
    done

    idx=0
    for ifname in $(get_network_interfaces_1g); do
        eval "ETH_1G_${idx}=${ifname}"
        (( idx+=1 ))
    done
}

# Get a list of the active network ports on this machine. The result is returned as an array of packs stored in the
# variable passed to the function.
#
# Options:
#  -l Only include listening ports
#
# For example:
# declare -A ports
# get_listening_ports ports
# einfo $(lval %ports[5])
# >> ports[5]=([proto]="tcp" [recvq]="0" [sendq]="0" [local_addr]="0.0.0.0" [local_port]="22" [remote_addr]="0.0.0.0" [remote_port]="0" [state]="LISTEN" [pid]="9278" [prog]="sshd" )
# einfo $(lval %ports[42])
# ports[42]=([proto]="tcp" [recvq]="0" [sendq]="0" [local_addr]="172.17.5.208" [local_port]="48899" [remote_addr]="173.194.115.70" [remote_port]="443" [state]="ESTABLISHED" [pid]="28073" [prog]="chrome" )
#
get_network_ports()
{
    $(opt_parse \
        "+listening l | Only include listening ports" \
        "__ports_list")

    local idx=0
    local first=1
    while read line; do

        # Expected netstat format:
        #  Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
        #  tcp        0      0 10.30.65.166:4013       0.0.0.0:*               LISTEN      42004/sfapp
        #  tcp        0      0 10.30.65.166:4014       0.0.0.0:*               LISTEN      42002/sfapp
        #  tcp        0      0 10.30.65.166:8080       0.0.0.0:*               LISTEN      42013/sfapp
        #  tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      19221/sshd
        #  tcp        0      0 0.0.0.0:442             0.0.0.0:*               LISTEN      13159/sfconfig
        #  tcp        0      0 172.30.65.166:2222      192.168.138.137:35198   ESTABLISHED 6112/sshd: root@not
        # ...
        #  udp        0      0 0.0.0.0:123             0.0.0.0:*                           45883/ntpd
        #  udp        0      0 0.0.0.0:161             0.0.0.0:*                           39714/snmpd
        #  udp        0      0 0.0.0.0:514             0.0.0.0:*                           39746/rsyslogd
        #
        # If netstat cannot determine the program that is listening on that port (not enough permissions) it will substitute a "-":
        #  tcp        0      0 0.0.0.0:902             0.0.0.0:*               LISTEN      -
        #  udp        0      0 0.0.0.0:43481           0.0.0.0:*                           -
        #

        # Compare first line to make sure fields are what we expect
        if [[ ${first} -eq 1 ]]; then
            local expected_fields="Proto Recv-Q Send-Q Local Address Foreign Address State PID/Program name"
            assert_eq "${expected_fields}" "${line}"
            first=0
            continue
        fi

        # Convert the line into an array for easy access to the fields
        # Replace * with 0 so that we don't get a glob pattern and end up with an array full of filenames from the local directory
        local fields
        array_init fields "$(echo ${line} | tr '*' '0')" " :/"

        # Skip this line if this is not TCP or UDP
        [[ ${fields[0]} =~ (tcp|udp) ]] || continue

        # Skip this line if the -l flag was passed in and this is not a listening port
        [[ ${listening} -eq 1 && ${fields[0]} == "tcp" && ! ${fields[7]} =~ "LISTEN" ]] && continue

        # If there is a - in the line, then netstat could not determine the program listening on this port.
        # Remove the - and add empty strings for the last two fields (PID and program name)
        if [[ ${line} =~ "-" ]]; then
            array_remove fields "-"
            fields+=("")
            fields+=("")
        fi

        # If this is a UDP port, insert an empty string into the "state" field
        if [[ ${fields[0]} == "udp" ]]; then
            fields[9]=${fields[8]}
            fields[8]=${fields[7]}
            fields[7]=""
        fi

        pack_set ${__ports_list}[${idx}] \
            proto=${fields[0]} \
            recvq=${fields[1]} \
            sendq=${fields[2]} \
            local_addr=${fields[3]} \
            local_port=${fields[4]} \
            remote_addr=${fields[5]} \
            remote_port=${fields[6]} \
            state=${fields[7]} \
            pid=${fields[8]} \
            prog=${fields[9]}

        (( idx += 1 ))

    done <<< "$(netstat --all --program --numeric --protocol=inet 2>/dev/null | sed '1d' | tr -s ' ')"
}
